src/catalog/src/durable/impls/persist.rs

```diff
@@ -486,6 +522,54 @@ impl PersistHandle {
     async fn get_user_version(&mut self, upper: Timestamp) -> Option<u64> {
         self.get_config(USER_VERSION_KEY, upper).await
     }
+
+    /// Appends `updates` to the catalog state and downgrades the catalog's upper to `next_upper`
+    /// iff the current global upper of the catalog is `current_upper`.
+    async fn compare_and_append(
+        &mut self,
+        updates: Vec<StateUpdate>,
+        current_upper: Timestamp,
+        next_upper: Timestamp,
+    ) -> Result<(), CatalogError> {
+        let mut batch_builder = self
+            .write_handle
+            .builder(Antichain::from_elem(current_upper));
+        for StateUpdate { kind, ts, diff } in updates {
+            batch_builder
+                .add(&kind, &(), &ts, &diff)
+                .await
+                .expect("invalid usage");
+        }
+        let mut batch = batch_builder
+            .finish(Antichain::from_elem(next_upper))
+            .await
+            .expect("invalid usage");
+        match self
+            .write_handle
+            .compare_and_append_batch(
```

Comment:
optional, especially since this is just code movement: You could save yourself a bit of verbosity here if you like by calling `compare_and_append` (which takes an iterator of updates) instead of `compare_and_append_batch`

Reply:
Oh nice, just updated, I think it looks a lot cleaner.